prober/derp.go

```diff
@@ -577,6 +580,196 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
 	return nil
 }
 
+// derpProbeBandwidthMu ensures that derpProbeBandwidthTUN doesn't run concurrently.
+// This is necessary to avoid conflicts trying to create the TUN device.
+var derpProbeBandwidthMu sync.Mutex
+
+// derpProbeBandwidthTUN takes two DERP clients (fromc and toc) connected to two
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another over TUN.
+func derpProbeBandwidthTUN(ctx context.Context, transferTime *expvar.Float, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64) error {
+	derpProbeBandwidthMu.Lock()
+	defer derpProbeBandwidthMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP.
+	dev, _, err := tstun.New(log.Printf, "utun9")
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer dev.Close()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	r, err := router.New(log.Printf, dev, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create router: %w", err)
+	}
+	defer r.Close()
+
+	err = r.Set(&router.Config{
+		LocalAddrs: []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+		Routes:     []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to set routes: %w", err)
+	}
+
+	// Read packets from TUN device
+	tunReadC := make(chan error, 1)
+	go func() {
+		// There's not a lot of thought in the below number. 5 doesn't use a ton of memory,
+		// and a bigger number like 20 didn't seem to increase throughput.
+		numBufs := 5
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tstun.PacketStartOffset))
```

Comment:
`tstun.PacketStartOffset` is WireGuard's transport header size, which is reserved space that is populated during encryption. Since we aren't doing crypto here it's irrelevant in the read direction, and can be removed. In the write direction it gets repurposed as scratch space for segmentation offload metadata (virtio_net_hdr).

Reply:
On Darwin, we need this to be at least [4 bytes](https://github.com/tailscale/wireguard-go/blob/tailscale/tun/tun_darwin.go#L228). We'll change to just 4 bytes.

Reply:
Turns out we need more than that, back to using `PacketStartOffset`.